platforms/common/include/px4_platform_common/i2c_spi_buses.h

```diff
@@ -59,31 +60,68 @@ enum class I2CSPIBusOption : uint8_t {
 class I2CSPIInstance
 {
 public:
-	I2CSPIInstance(I2CSPIBusOption bus_option, int bus, int type)
-		: _bus_option(bus_option), _bus(bus), _type(type) {}
-
 	virtual ~I2CSPIInstance() = default;
 
 private:
+	I2CSPIInstance(I2CSPIBusOption bus_option, int bus, uint8_t i2c_address, uint16_t type)
+		: _bus_option(bus_option), _bus(bus), _type(type), _i2c_address(i2c_address) {}
+
 	friend class BusInstanceIterator;
 	friend class I2CSPIDriverBase;
 
 	const I2CSPIBusOption _bus_option;
 	const int _bus;
-	const int _type; ///< device type (driver-specific)
+	const int16_t _type; ///< device type (driver-specific)
+	const int8_t _i2c_address; ///< I2C address (optional)
 };
 
-struct BusCLIArguments {
+class BusCLIArguments
+{
+public:
+	BusCLIArguments(bool i2c_support, bool spi_support)
+		: _i2c_support(i2c_support), _spi_support(spi_support) {}
+
+	/**
+	 * Parse CLI arguments (for drivers that don't need any custom arguments, otherwise getopt() should be used)
+	 * @return command (e.g. "start") or nullptr on error or unknown argument
+	 */
+	const char *parseDefaultArguments(int argc, char *argv[]);
+
+	/**
+	 * Like px4_getopt(), but adds and handles i2c/spi driver-specific arguments
+	 */
+	int getopt(int argc, char *argv[], const char *options);
+
+	/**
+	 * returns the current optional argument (for options like 'T:'), or the command (e.g. "start")
+	 * @return nullptr or argument/command
+	 */
+	const char *optarg() const { return _optarg; }
+
+
 	I2CSPIBusOption bus_option{I2CSPIBusOption::All};
-	int type{0}; ///< device type (driver-specific)
+	uint16_t type{0}; ///< device type (driver-specific)
 	int requested_bus{-1};
 	int chipselect_index{1};
 	Rotation rotation{ROTATION_NONE};
 	int bus_frequency{0};
-	int spi_mode{-1};
+	spi_mode_e spi_mode{SPIDEV_MODE3};
+	uint8_t i2c_address{0}; ///< optional I2C address: a driver can set this to allow configuring the I2C address
+
 	int custom1{0}; ///< driver-specific custom argument
 	int custom2{0}; ///< driver-specific custom argument
 	void *custom_data{nullptr}; ///< driver-specific custom argument
+
+	// driver defaults, if not specified via CLI
+	int default_spi_frequency{20 * 1000 * 1000}; ///< default spi bus frequency [Hz]
```

Comment:
1 MHz SPI and 100 kHz I2C might be safer defaults

Reply:
What is the ripple effect if we inadvertently set lower frequencies without noticing on a bunch of boards? I've been there, done that and people didn't realize it until very late into the investigation. 20 MHz is high, but pretty much any SPI device takes 10 MHz.

Reply:
I was going for what most drivers do, so that we don't have the problems Lorenz mentions.
We can also have no defaults and enforce all drivers to have to set their values.
Any preferences?

Reply:
At the moment forcing all drivers their value seems appropriate.

These days we have quite a few sensors that operate below 10 MHz. https://docs.google.com/spreadsheets/d/1OUOSr8TDeFwcNdao5b00iQvMUCQIgpeUJf85VwGpIC0/edit#gid=0

Reply:
Done in https://github.com/PX4/Firmware/pull/14386/commits/45c9899267d86a5204c4c00d96389de020932052